src/diffusers/pipelines/pipeline_utils.py

```diff
@@ -762,6 +801,25 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         # pop out "_ignore_files" as it is only needed for download
         config_dict.pop("_ignore_files", None)
 
+        if dduf:
+            import tarfile
+
+            tar_file_path = os.path.join(cached_folder, dduf)
+            extract_to = os.path.join(cached_folder, f"{dduf}_extracted")
+            # if tar file, we need to extract the tarfile and remove it
+            if os.path.isfile(tar_file_path):
+                if tarfile.is_tarfile(tar_file_path):
+                    with tarfile.open(tar_file_path, "r") as tar:
+                        tar.extractall(extract_to)
+                    # remove tar archive to free memory
+                    os.remove(tar_file_path)
+                    # rename folder to match the name of the dduf archive
+                    os.rename(extract_to, tar_file_path)
```

Comment:
(Moving my comment to the right place, sorry)

I think that expanding the archive and removing it could be problematic.

- If I understand correctly, we would be removing the symlink to the blob, but not the blob itself. We won't be saving disk space.
- If we did remove the blob, I think it'd trigger another download next time we call `.from_pretrained`.
- Removing the reference could maybe have consequences when checking for newer versions in the Hub (not sure).

When we change to using zips, I wonder if we could extract files progressively in memory during the load process, and keeping the big .dduf file as is. Having a disk snapshot is convenient, but I can see no easy solution to avoid duplicating space.

Reply:
I agree. I will check if it is possible to stream directly from the zip or at least extract progressively in memory as you said

Reply:
Thanks for the link ! I'll have a look !

Reply:
BTW on the Hub side we want to constraint the types of Zip a lot:
- no compression only (to make sure it's deduplicated well by XetHub)
- single disk (internal simplifications of Zip)
- no encryption etc

Reply:
To ensure restrictions are met, we'll provide tooling in Python to save/load DDUF files in a consistent way: https://github.com/huggingface/huggingface_hub/pull/2692. This will be part of `huggingface_hub` to that it can be easily reused by any library without having necessarily `diffusers` as a dependency.